### PR TITLE
[doc] Improve wibox.widget.textclock format documentation

### PR DESCRIPTION
**Pull request by Aire-One**
Hello,

I initially started this work in the scope of #3220.

Textclock is an interesting widget as it inherits from another widget: textbox. I think the doc should make it clear, and we can play with the new ldoc tags added by #2924 in this effect.

I'm not happy with how commit 4f64f80 turned out, tho. It involved a big copy-paste of the textbox documentation to `docs/common/wibox_widget_textbox.ldoc`. I'm afraid it will not be maintained when textbox will be updated...

Commit 99866f4 is a bit off-topic but still deeply related to this inheritance subject. I think we can have a new section to present objects hierarchy diagram to the user. It makes the inheritances relations clear, and we can benefit from clarity here.
The current implementation for the diagram relies on HTML nested list. I think we could add more here. Maybe a "known subclasses" list... Anyway, I'm open to feedback on it :) 

object-hierarchy

Funny note: `tests/examples/wibox/widget/defaults/textclock.lua` instantiates a textclock widget to generate its image, so it will display the date and time the doc is built.

### CODE REVIEW

File: lib/wibox/widget/textclock.lua
Comment (ShayAgros) [inline lib/wibox/widget/textclock.lua:37]: Need to add space after the double dash. Otherwise the link below (g-date-time-format) isn't rendered properly

**Comment by Elv13** [comment]
Sorry for the delay. Awesome has been rather low on my COVID-era priority list. Do you think this is ready to merge once @ShayAgros comments are addressed?

Also, do you think we could auto-generate this using  a `@something` ldoc tag in the header rather than using the examples framework to render the markdown? This would be easier to apply to the entire doc rather than having to make an example per class. However it makes the already gigantic `config.ld` and html template larger. Still, I think it would be cleaner and be rather easy to convert your current code into config.ld/template

**Comment by ShayAgros** [comment]
When running `make` and opening `build/doc/widgets/wibox.widget.textbox.html` I see the addition of the examples there (nicely done btw), but don't see the inheritance diagram you showed at the pr header. Is it expected ? Tried to dig into it myself, but haven't pinned down the reason yet.

File: lib/wibox/widget/textclock.lua
Comment (Aire-One) [reply lib/wibox/widget/textclock.lua:37]: Nope, with the additional space, the generated documentation will have the double dash drawn before the table. If you check how other files from `docs/common/*.ldoc` are included, they do the same.

It's a tricky hack to integrate the file content without making it a "comment of a comment" (output `----` vs `-- --`). ldoc will interpret `----` as a regular comment opening while `-- --` will be a comment followed by a double dash (to be rendered).

You can check that after running `make ldoc` by inspecting the `build/lib/wibox/widget/textclock.lua` file and checking the generated code that ldoc will parse to generate the API references.

File: lib/wibox/widget/textclock.lua
Comment (Aire-One) [reply lib/wibox/widget/textclock.lua:4]: The full sentence says `Awesome WM's wiboxes system` with a link to the documentation at https://awesomewm.org/apidoc/documentation/03-declarative-layout.md.html.

Still, "wiboxes system" is probably an unnecessarily complicated way to says it's a widget from Awesome library...

I'll go with `Awesome WM's widgets system` for now. The objective of this sentence was to help visualize where the class is on the internal API conception. I think this could be removed once I'll have finalized the "inheritance diagram" stuff. I guess a good diagram will be better than a strange sentence...

**Comment by Aire-One** [comment]
I have split this PR to make the "improve textclock doc" a first ready to merge work. Thank you, @ShayAgros, for your review!

I'll work on a new branch to make progress on the inheritance diagram.

(The work I previously submitted is still available in a backup branch at https://github.com/Aire-One/awesome/tree/doc_textclock_backup)

> Also, do you think we could auto-generate this using a @something ldoc tag in the header rather than using the examples framework to render the markdown? This would be easier to apply to the entire doc rather than having to make an example per class. However it makes the already gigantic config.ld and html template larger. Still, I think it would be cleaner and be rather easy to convert your current code into config.ld/template

Using a ldoc tag is definitely something I'd like to explore. I was not sure if it was the right way to generate html content at the "class/module" level... Anyway, I'll try to see how far we can go with ldoc tags, maybe I see too far right now, but I hope we can use ldoc to automatically generate an inheritance tree we can use to render the inheritance diagram and even include inherited properties!

File: lib/wibox/widget/textclock.lua
Comment (ShayAgros) [reply lib/wibox/widget/textclock.lua:35]: Thanks for this detail explanation, I wasn't aware of `build/lib/wibox/widget/textclock.lua` file (this simplifies my code investigation quite a lot).  Running `make ldoc` on the following snippet:
```
----<table class='widget_list' border=1>
-- </table>
--
-- You can read more on the GLib DateTime format at
-- [https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-format](https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-format).
```

Causes the 'html' output `build/doc/widgets/wibox.widget.textclock.html` to show the `https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-format` part as simple text rather than a link. Looking at the html of this file I see:
```
 You can read more on the GLib DateTime format at
 [https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-format](https://developer.gnome.org/glib/stable/glib-GDateTime.html#g-date-time-format).
```
(so `ldoc` didn't render the markdown at all)

Adding space to the `<table>` line:
```
---- <table class='widget_list' border=1>
```

make the link being rendered correctly. I don't know how to solve this problem atm (I did miss the extra dashes added by my proposed solution), but I'd look into it.

Is the link, in this commit, rendered correctly on your setup ? (maybe I have a version issue, as I often have with Ubutnu)

**Comment by Aire-One** [comment]
Fixes #3220

**Comment by Aire-One** [comment]
Closes #3220

**Review by Elv13 — DISMISSED** [review]

**Review by actionless — DISMISSED** [review]

**Comment by Aire-One** [comment]
(close/reopen to trigger Travis. We already have 2 approvals, so I hope PR will be auto merged when CI will be completed)

**Comment by Aire-One** [comment]
> I hope PR will be auto merged when CI will be completed

Unless Travis failed... It seems to 5th job was "canceled". I'm not sure what happened.

**Comment by Elv13** [comment]
> Unless Travis failed... It seems to 5th job was "canceled". I'm not sure what happened.

I restarted it a few time, it doesn't want to finish. I will need a debug this, so I will keep this unmerged so I can reproduce. What happens is that `set -x` seems to kill the script for unknown reason. Something is returning a non-zero ret code while `luacheck` is printing stuff. The logs doesn't make it obvious about *what* is causing this. It's rather similar to the bug that prevented the API doc to be generated 95% of the time in the last couple weeks. If I restart it enough time, it will eventually pass, but I would like to know why it fails.

**Comment by Elv13** [comment]
Looks like there's 2 issues. One seems to be a Travis-ci regression where the log is just cropped. I added a couple `trap` and they are not visibly triggerred. Plus the fact that the output cuts mid-line is also highly suspicious.

But beside, the problem was that one of the URL in the doc tripped the line length limit. So that's what "really" made the build fail, despite Travis not giving any indication of that being the case. I rebased the PR with a fix.

**Review by Elv13 — APPROVED** [review]

**Review by actionless — APPROVED** [review]

**Comment by actionless** [comment]
im also fine leaving debug stuff in travis, as in other PR he gonna be removed anyway

**Pull request merged**
